src/catchup/IndexBucketsWork.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "IndexBucketsWork.h"
+#include "bucket/BucketIndex.h"
+#include "bucket/BucketList.h"
+#include "bucket/BucketManager.h"
+#include "util/HashOfHash.h"
+#include "util/UnorderedSet.h"
+#include "util/types.h"
+#include "work/WorkWithCallback.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+IndexBucketsWork::IndexWork::IndexWork(Application& app,
+                                       std::shared_ptr<Bucket> b)
+    : BasicWork(app, "index-work", BasicWork::RETRY_NEVER), mBucket(b)
+{
+}
+
+BasicWork::State
+IndexBucketsWork::IndexWork::onRun()
+{
+    if (mDone)
+    {
+        return State::WORK_SUCCESS;
+    }
+
+    postWork();
+    return State::WORK_WAITING;
+}
+
+bool
+IndexBucketsWork::IndexWork::onAbort()
+{
+    return true;
+};
+
+void
+IndexBucketsWork::IndexWork::postWork()
+{
+    Application& app = this->mApp;
+
+    std::weak_ptr<IndexWork> weak(
+        std::static_pointer_cast<IndexWork>(shared_from_this()));
+    app.postOnBackgroundThread(
+        [&app, &b = mBucket, weak]() {
+            auto self = weak.lock();
+            if (!self || self->isAborting())
+            {
+                return;
+            }
+
+            b->setIndex(BucketIndex::createIndex(app.getBucketManager(),
```

Comment:
There's a check on line 51 that exits early if work is aborting, so it doesn't seem like we need to check that here

Reply:
I think it still serves some purpose. Creating the index could take a significant amount of time (minutes), so it's very possible that the Work was not aborting when it started indexing but is now. In practice, it doesn't matter because we return null indexes if we detect the system is aborting, but still probably good defensive practice to not start indexing if you are aborting and to not set the index if you are aborting.

Reply:
oh I see, yes you both are right, I missed that there are two actions happening on this line